feat: Add OpenTelemetrySpanExt methods for direct span event creation

alessandrobologna opened the pull request:
This PR adds methods to `OpenTelemetrySpanExt` for directly adding OpenTelemetry span events, addressing the limitations of using `tracing::event!` for events with dynamic attributes.

Fixes #200

## Motivation
As discussed in issue #200, the current `tracing::event!` macro requires attribute keys to be known at compile time. This makes it difficult or impossible to record OpenTelemetry span events with dynamic attributes (e.g., from processing arbitrary JSON payloads) without resorting to lossy serialization, which hinders observability.

This change provides a way to directly add `opentelemetry::trace::Event` objects with dynamic `Vec<KeyValue>` attributes, similar to how `set_attribute` allows setting dynamic span attributes. This improves flexibility, makes porting code easier, and aligns better with the capabilities of the OpenTelemetry Span API.

## Solution
This PR extends the `OpenTelemetrySpanExt` trait with two new methods:

1.  `add_otel_span_event<T>(&self, name: T, attributes: Vec<KeyValue>) where T: Into<Cow<'static, str>>`: Adds an OpenTelemetry span event with the current timestamp.
2.  `add_otel_span_event_with_timestamp<T>(&self, name: T, timestamp: SystemTime, attributes: Vec<KeyValue>) where T: Into<Cow<'static, str>>`: Adds an OpenTelemetry span event with a specified timestamp.

The implementation uses the existing `with_subscriber` -> `downcast_ref::<WithContext>` -> `with_context` pattern found in other `OpenTelemetrySpanExt` methods (`set_attribute`, `set_status`) to access the internal `OtelData` builder and push the event.

The name `add_otel_span_event` was chosen over `add_event` to avoid potential confusion with the `tracing::event!` macro within the context of this crate.

This PR includes:
- The trait method definitions and implementations in `src/span_ext.rs`.
- Corresponding test cases in `tests/span_ext.rs`.
- Updates to `README.md` documentation.
- A new example file `examples/opentelemetry-span-event.rs`.
- Updated `CHANGELOG.md`.

alessandrobologna commented:
Hi @djc (and other maintainers),
Just wanted to add a couple of notes about this PR:
Clippy Fix Commit: You'll notice the last commit (c54ba96) addresses a Clippy lint (`double_ended_iterator_last` in `tests/metrics_publishing.rs`). This warning was surfaced by the CI checks (-D warnings) and appeared unrelated to the core feature addition, so I kept it as a separate commit for clarity.
Versioning & Changelog: I haven't bumped the crate version in Cargo.toml, and the changelog entries for the new feature and example are currently under the ## Unreleased section in CHANGELOG.md.

I'm happy to handle the version bump and finalize the changelog entry if after your review this PR is ready to merge. I wasn't sure about the appropriate version target for this addition, since it's adding new API surface (non-breaking) to a 0.x crate, would you typically target a minor version bump (e.g., 0.31.0) or a patch release (e.g., 0.30.1)? 
Happy to follow whatever the project's policy is.
Thanks for taking a look!

djc commented on src/span_ext.rs:
````diff
@@ -152,6 +157,63 @@ pub trait OpenTelemetrySpanExt {
+    ///
+    /// app_root.add_otel_span_event("job_completed", vec![KeyValue::new("status", "success")]);
+    /// ```
+    fn add_otel_span_event<T>(&self, name: T, attributes: Vec<KeyValue>)
````
Are there other kinds of events in OTel? The name is a little verbose, so would be nice if we can get away with something a bit shorter like `add_otel_event()`.

alessandrobologna replied:
Yes, there's a new [WIP spec for OTel Events](https://opentelemetry.io/docs/specs/semconv/general/events/), as a new signal type. More background about it in here: 

- https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/0202-events-and-logs-api.md
- https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/0265-event-vision.md
- https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/4430-span-event-api-deprecation-plan.md

I think though that it's still far away from being widely implemented and adopted, so I am happy with the span events API for now (and I don't necessarily agree that it was a good idea to introduce a new signal type, but that's digressing).

I agree that the name is verbose, but `add_otel_event` would then be a little ambiguous when those specs are more widely adopted and known. Alternatively, we could use `add_event` (as in the OTel Span trait, in the same way that you have `add_link`) and since this is a method of a trait that is meant to "wrap" an OTel span, maybe it would be clear that is not a Tracing event. Let me know your preference and I will update the PR.

djc commented on src/span_ext.rs:
````diff
@@ -152,6 +157,63 @@ pub trait OpenTelemetrySpanExt {
+    ///
+    /// app_root.add_otel_span_event("job_completed", vec![KeyValue::new("status", "success")]);
+    /// ```
+    fn add_otel_span_event<T>(&self, name: T, attributes: Vec<KeyValue>)
````
Nit: please use impl trait in arguments consistently where possible, so this should have `name: impl Into<Cow<'static, str>>` (everywhere).

djc commented on examples/opentelemetry-span-event.rs:
```diff
@@ -0,0 +1,65 @@
+use opentelemetry::trace::TracerProvider as _;
```
This seems a little niche so I'd prefer to avoid maintaining an example for it.

alessandrobologna replied:
Ok I can remove it

djc commented on src/span_ext.rs:
```diff
@@ -237,4 +299,33 @@ impl OpenTelemetrySpanExt for tracing::Span {
+            let mut event = Some(opentelemetry::trace::Event::new(
+                name, timestamp, attributes, 0,
+            ));
+            if let Some(get_context) = subscriber.downcast_ref::<WithContext>() {
+                get_context.with_context(subscriber, id, move |data, _tracer| {
+                    if let Some(event) = event.take() {
+                        data.builder.events.get_or_insert_with(Vec::new).push(event);
+                    }
+                });
+            }
```
Nit: I'm aware this is mostly copy/pasting, but please use let-else to reduce rightward drift where possible. (If you want to fix up the original function too, that would be nice, but please keep it in a separate commit.)

djc commented on CHANGELOG.md:
```diff
@@ -1,5 +1,13 @@
 
+### Added
+
+- Add `OpenTelemetrySpanExt::add_otel_span_event` function to allow adding OpenTelemetry
+  events directly to a `tracing::Span`, enabling the use of dynamic attribute keys.
+- Add `OpenTelemetrySpanExt::add_otel_span_event_with_timestamp` function, similar to
+  `add_otel_span_event`, but allowing a specific `SystemTime` to be provided for the event.
```
Nit: I think these changes belong in a single bullet point.

djc commented on CHANGELOG.md:
```diff
@@ -1,5 +1,13 @@
+  events directly to a `tracing::Span`, enabling the use of dynamic attribute keys.
+- Add `OpenTelemetrySpanExt::add_otel_span_event_with_timestamp` function, similar to
+  `add_otel_span_event`, but allowing a specific `SystemTime` to be provided for the event.
+- Add `examples/opentelemetry-span-event.rs` demonstrating the usage of the new span event methods.
```
Don't forget to remove this when you remove the example.

alessandrobologna replied:
will do

djc reviewed (commented):
This is mostly looking good, thanks!

Please keep the clippy commits separate and squash any additional changes into your original commit (might be easier to reorder them so that the clippy commits come first).

alessandrobologna commented:
@djc I think I have addressed all the changes requested. Let me know if you spot something else.

djc commented:
As mentioned:

> squash any additional changes into your original commit

Please squash the last two commits ("rename span event methods" and "use impl trait consistently"), which AFAICT only touch new code you have added, into your original commit ("add OpenTelemetrySpanExt methods").

alessandrobologna commented:
Hi @djc, thanks! I think I've addressed the feedback:

* Kept the clippy fixes separate and moved them first.
* Squashed the feature addition and subsequent refactorings into the final commit (78e8247).

The commit history should now match your suggestion. Let me know if there's something else!

djc commented:
You actually didn't address my feedback exactly, because you also squashed the refactoring that refactored other pre-existing code that you didn't add, so that's a little yucky from a reviewing perspective. However, I don't think it's worth doing another round trip on this, so I'll just merge this.

alessandrobologna commented:
Apologies, I misunderstood, I just thought you wanted the clippy fixes separated. I see your point now.
